app/controllers/uploads_controller.rb

```diff
@@ -12,22 +10,29 @@ def new
     @upload = new_upload(params[:csv_type])
     return if @upload.csv_type_check?
 
-    log_and_display_error(errors_for_alert([@upload]), 'Warning')
+    alert_and_log(@upload.errors.full_messages.join(', '))
     redirect_to dashboards_path
   end
 
   def create
     @upload = Upload.create(merged_params)
 
     begin
-      data = load_csv
-      display_failed_instances(data.failed_instances)
+      failed = load_csv.failed_instances
```

Comment:
This block feels a little confusing to me. You get half of your warnings from the load_csv result object, half from the upload object. I'm not suggesting a major refactor at this point but the object encapsulation feels a little off.